drivers/can/can_esp32_twai.c

```diff
@@ -18,12 +19,42 @@
 
 LOG_MODULE_REGISTER(can_esp32_twai, CONFIG_CAN_LOG_LEVEL);
 
+/*
+ * Newer ESP32-series MCUs like ESP32-C3 and ESP32-S2 have some slightly different registers
+ * compared to the original ESP32, which is fully compatible with the SJA1000 controller.
+ *
+ * The names with TWAI_ prefixes from Espressif reference manuals are used for these incompatible
+ * registers. Compatible registers use the original SJA1000 names.
+ */
+#ifndef CONFIG_SOC_ESP32
+
+/* TWAI_BUS_TIMING_0_REG is incompatible with CAN_SJA1000_BTR0 */
+#define TWAI_BAUD_PRESC_MASK            GENMASK(12, 0)
+#define TWAI_SYNC_JUMP_WIDTH_MASK       GENMASK(15, 14)
+#define TWAI_BAUD_PRESC_PREP(brp)	FIELD_PREP(TWAI_BAUD_PRESC_MASK, brp)
+#define TWAI_SYNC_JUMP_WIDTH_PREP(sjw)	FIELD_PREP(TWAI_SYNC_JUMP_WIDTH_MASK, sjw)
+
+/* TWAI_CLOCK_DIVIDER_REG is incompatible with CAN_SJA1000_CDR */
+#define TWAI_CD_MASK			GENMASK(7, 0)
+#define TWAI_CD_PREP(brp)		FIELD_PREP(TWAI_CD_MASK, brp)
+#define TWAI_CLOCK_OFF			BIT(8)
+
+/*
+ * Further incompatible registers currently not used by the driver:
+ * - TWAI_STATUS_REG has new bit 8: TWAI_MISS_ST
+ * - TWAI_INT_RAW_REG has new bit 8: TWAI_BUS_STATE_INT_ST
+ * - TWAI_INT_ENA_REG has new bit 8: TWAI_BUS_STATE_INT_ENA
+ */
+#endif /* !CONFIG_SOC_ESP32 */
+
 struct can_esp32_twai_config {
 	mm_reg_t base;
 	const struct pinctrl_dev_config *pcfg;
 	const struct device *clock_dev;
 	const clock_control_subsys_t clock_subsys;
 	int irq_source;
+	/* 32-bit variant of output clock divider register required for non-ESP32 MCUs */
+	uint32_t cdr32;
```

Comment:
I know, this *could* also be wrapped in an `#ifndef CONFIG_SOC_ESP32`, but I decided against it because it would make the macros at the bottom of the file much harder to read. 4 bytes of wasted RAM should be acceptable.

Reply:
If you have a good idea how the `.cdr32 = CAN_ESP32_TWAI_DT_CDR_INST_GET(inst),` assignment can be excluded without duplicating the entire macro, let me know.

Reply:
Take a look at how the `FLEXCAN_PINCTRL_INIT` macro is for conditional config struct member assignment here: https://github.com/zephyrproject-rtos/zephyr/blob/147d9cb1272b39e119025953acdb3d19ba52e45e/drivers/can/can_mcux_flexcan.c#L953